### PR TITLE
Feature: Basic support for Google Artifact Registry

### PR DESCRIPTION
**Pull request by patricklubach**
Hello there,

as the current version of terraform-google-gcr-cleaner only supports Google Container Registry, I wanted to add basic support for Google Artifact Registry as the gcr-cleaner already supports it and I need this feature for myself. Please have a look on this.

### CODE REVIEW

**Comment by patricklubach** [comment]
I have to run git-commit and extend the README. I will add the additional variable shortly.

**Comment by patricklubach** [comment]
Hello @anouarchattouna, @PascalBourdier and @mirakl-admin, do anyone of have time to have a look on this? Would highly appreciate if you catch time.

**Comment by anouarchattouna** [comment]
Hi @patricklubach, thank you for this PR.
We will have a look ASAP!

File: locals.tf [hunk 1]
```diff
@@ -64,6 +64,20 @@ locals {
     ] if gcr.repositories != null
   ])
 
+  # create gar_repositories list and sets default values for optional fields.
+  gar_repositories = [
+    for gar in var.gar_repositories : {
+      repo           = "${gar.region}-docker.pkg.dev/${gar.project_id}/${gar.name}"
```
Comment (anouarchattouna) [inline locals.tf:70]: It looks like the region is required here, but what about retrieving the project_id from the provider itself when not provided?

File: variables.tf [hunk 2]
```diff
@@ -117,6 +117,40 @@ EOF
...
+        tag_filter     = (Deprecated) If specified, any image where the first tag matches this given regular expression will be deleted. The image will not be deleted if other tags match the regular expression (optional(string))
+        tag_filter_any = If specified, any image with at least one tag that matches this given regular expression will be deleted. The image will be deleted even if it has other tags that do not match the given regular expression (optional(string))
+        tag_filter_all = If specified, any image where all tags match this given regular expression will be deleted. The image will not be delete if it has other tags that do not match the given regular expression (optional(string))
+    })))
+}))
+```
+EOF
+
+  type = list(object({
+    project_id = string
```
Comment (anouarchattouna) [inline variables.tf:140]: Why this is not optional like in `gcr_repositories`?

File: iam.tf [hunk 3]
```diff
@@ -36,3 +36,16 @@ resource "google_project_iam_member" "this" {
   role    = "roles/browser"
   member  = "serviceAccount:${google_service_account.cleaner.email}"
 }
+
+# Grant cleaner service account roles/artifactregistry.writer role to read the repository
+# and delete container images.
+resource "google_artifact_registry_repository_iam_member" "member" {
+  provider = google-beta
```
Comment (anouarchattouna) [inline iam.tf:43]:
```suggestion
```

File: iam.tf [hunk 4]
```diff
@@ -36,3 +36,16 @@ resource "google_project_iam_member" "this" {
...
+  for_each = { for repo in var.gar_repositories : repo.name => repo }
```
Comment (anouarchattouna) [inline iam.tf:44]:
```suggestion
  for_each = { 
    for repo in var.gar_repositories : repo.name => repo 
  }
```

File: iam.tf [hunk 5]
```diff
@@ -36,3 +36,16 @@ resource "google_project_iam_member" "this" {
...
+
+  project    = each.value.project_id
+  location   = each.value.region
+  repository = "projects/${each.value.project_id}/locations/${each.value.region}/repositories/${each.value.name}"
+  role       = "roles/artifactregistry.writer"
+  member     = "serviceAccount:${google_service_account.cleaner.email}"
+}
```
Comment (anouarchattouna) [inline iam.tf:51]:
```suggestion

  provider = google-beta
}
```

**Review by anouarchattouna — COMMENTED** [review]
In addition, the `google-beta` provider should be added and configured in examples

**Comment by anouarchattouna** [comment]
I am facing this error when testing the minimal example
```console
│ Error: Error retrieving IAM policy for artifactregistry repository "projects/my-test-project/locations/europe-west3/repositories/nginx": googleapi: Error 403: The caller does not have permission
│
│   with module.gcr_cleaner.google_artifact_registry_repository_iam_member.member["nginx"],
│   on ../../iam.tf line 42, in resource "google_artifact_registry_repository_iam_member" "member":
│   42: resource "google_artifact_registry_repository_iam_member" "member" {
```

**Comment by patricklubach** [comment]
I'll try to quickly resolve all your findings. Thank you very much for reviewing!

File: .github/CODE_OF_CONDUCT.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
+# Contributor Covenant Code of Conduct
```
Comment (anouarchattouna) [inline .github/CODE_OF_CONDUCT.md:1]: IMO, this kind of file should not be added to your PR.

Comment (patricklubach) [reply .github/CODE_OF_CONDUCT.md:1]: TBH, I don't know why this is included. All I did, is rebased my branch with the current main branch.

**Comment by patricklubach** [comment]
> I am facing this error when testing the minimal example
> 
> ```
> │ Error: Error retrieving IAM policy for artifactregistry repository "projects/my-test-project/locations/europe-west3/repositories/nginx": googleapi: Error 403: The caller does not have permission
> │
> │   with module.gcr_cleaner.google_artifact_registry_repository_iam_member.member["nginx"],
> │   on ../../iam.tf line 42, in resource "google_artifact_registry_repository_iam_member" "member":
> │   42: resource "google_artifact_registry_repository_iam_member" "member" {
> ```

I tested minimal example and it works for me. No permissions is truly a very generic issue. Is there some way I can help you?

File: .github/CODE_OF_CONDUCT.md (hunk 6 above)
Comment (anouarchattouna) [reply .github/CODE_OF_CONDUCT.md:1]: Could you please rebase yours with the last upstream `main` branch?

Comment (patricklubach) [reply .github/CODE_OF_CONDUCT.md:1]: I did but seems to be no changes to this issue. What's weird, is when I am doing a `git diff upstream/main` locally, I cannot see that these files are changed.

**Comment by patricklubach** [comment]
@anouarchattouna I created another PR #39 that do not have all these file changes and close this one. I think the reason for this was a bit messed up git history. Sorry for that inconvenience.

**Pull request closed without merging**